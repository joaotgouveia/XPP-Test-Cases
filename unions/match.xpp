type my_union = union [ #first: int, #second: string ];
(
let f = fn a: my_union => {
    match a {
        | #first(b) -> b
        | #second(c) -> 1
    }
}; f(#second("123"))
);;

type my_union = union [ #first: int, #second: string ];
(
let f = fn a: my_union => {
    match a {
        | #first(b) -> b
        | #second(b) -> 10
        | #third(c) -> 123
    }
}; f(#first(321))
);;

type my_union = union [ #first: int, #second: string ];
(
let f = fn a: my_union => {
    match a {
        | #first(b) -> b
        | #third(c) -> 123
    }
}; f(#first(321))
);;

type my_union = union [ #first: int, #second: string ];
(
let f = fn a: my_union => {
    match a {
        | #first(b) -> "123"
        | #second(c) -> c
    }
}; f(#second("123"))
);;

type my_union = union [ #first: union [ #nested: string ], #second: string ];
(
let f = fn a: my_union => {
    match a {
        | #first(b) -> match b {
            | #nested(c) -> c
            }
        | #second(c) -> c
    }
}; f(#first(#nested("321")))
);;

type my_struct = struct { a: int, b: string };
type my_union = union [ #first: my_struct, #second: string ];
(
let f = fn a: my_union => {
    match a {
        | #first(b) -> b.a
        | #second(c) -> 123
    }
}; f(#first({ a = 1 ; b = "123" }))
);;

type my_union = union [ #Nil: () ];
(
let f = fn a: my_union => {
    match a {
        | #Nil -> 1
    }
}; 
f(#Nil())
);;

type hehe = union [ #Nil: () ];
type my_union = union [ #teste: hehe ];
(
let f = fn a: my_union => {
    match a {
        | #teste(a) -> match a {
            | #Nil -> 1
        }
    }
}; 
f(#teste(#Nil()))
);;
