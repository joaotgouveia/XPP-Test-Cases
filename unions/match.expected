Value: 1
Type: int

Value: 321
Type: int

throws TypeCheckError

Value: "123"
Type: string

Value: "321"
Type: string

Value: 1
Type: int

Value: 1
Type: int

Value: 1
Type: int
